[SPARK-5437] Fix DriverSuite and SparkSubmitSuite timeout issues

andrewor14 opened the pull request:
In DriverSuite, we currently set a timeout of 60 seconds. If after this time the process has not terminated, we leak the process because we never destroy it.

In SparkSubmitSuite, we currently do not have a timeout so the test can hang indefinitely.

Commit:
Stop SparkContext after every individual test

SparkQA commented:
[Test build #26181 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26181/consoleFull) for   PR 4230 at commit [`8092c36`](https://github.com/apache/spark/commit/8092c36831bb2b348f566bb4bd9a8d234cc5fc3d).
 * This patch merges cleanly.

SparkQA commented:
[Test build #26181 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26181/consoleFull) for   PR 4230 at commit [`8092c36`](https://github.com/apache/spark/commit/8092c36831bb2b348f566bb4bd9a8d234cc5fc3d).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class DenseMatrix(`

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26181/
Test PASSed.

Commit:
Fix timeout behaviors in both suites

This ensures two things: (1) both suites use timeouts, and (2)
both suites actually terminate the respective processes if
timeouts occur.

SparkQA commented:
[Test build #26188 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26188/consoleFull) for   PR 4230 at commit [`f5c80fd`](https://github.com/apache/spark/commit/f5c80fd67a3637d9868a9afec1dae9ac37044b16).
 * This patch merges cleanly.

SparkQA commented:
[Test build #26188 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/26188/consoleFull) for   PR 4230 at commit [`f5c80fd`](https://github.com/apache/spark/commit/f5c80fd67a3637d9868a9afec1dae9ac37044b16).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/26188/
Test PASSed.

vanzin commented on core/src/test/scala/org/apache/spark/DriverSuite.scala:
```diff
@@ -28,31 +28,30 @@ import org.apache.spark.util.Utils
 
 class DriverSuite extends FunSuite with Timeouts {
 
+  // Regression test for SPARK-530: "Spark driver process doesn't exit after finishing"
```
Since you're touching this, I kinda like the approach of adding the bug number to the `test()` call that I've seen in other places.

andrewor14 replied:
I just moved this line, but I can move it there too

vanzin commented:
LGTM.

andrewor14 commented:
Merging into master

squito commented on core/src/test/scala/org/apache/spark/DriverSuite.scala:
```diff
@@ -28,31 +28,30 @@ import org.apache.spark.util.Utils
+        Map("SPARK_TESTING" -> "1", "SPARK_HOME" -> sparkHome))
+      failAfter(60 seconds) { process.waitFor() }
+      // Ensure we still kill the process in case it timed out
+      process.destroy()
```
@andrewor14 shouldn't the `destroy` be in a `finally` block?  Otherwise the process isn't destroyed if you hit the timeout.

(I know this was committed a while back, and the test is even ignored now -- but I was looking at some flaky tests and this just happened to catch my eye, so I was curious)

andrewor14 replied:
yes, it would be good to fix it, though I don't think it will solve the flakiness that we currently encounter

andrewor14 replied:
by the way I'm addressing this at #6886
